cocos2d/Platforms/Mac/CCViewMacGL.m

```diff
@@ -0,0 +1,296 @@
+    // When resizing the view, -reshape is called automatically on the main thread
+    // Add a mutex around to avoid the threads accessing the context simultaneously when resizing
+    
+    [self lockOpenGLContext];
+    
+    NSRect rect = [self convertRectToBacking:self.bounds];
+    
+    CCDirector *director = _director;
+    [director reshapeProjection: NSSizeToCGSize(rect.size) ];
+    
+    // avoid flicker
+    // Only draw if there is something to draw, otherwise it actually creates a flicker of the current glClearColor
+    if(director.runningScene){
+        [director mainLoopBody];
+    }
+    //	[self setNeedsDisplay:YES];
+    
+    [self unlockOpenGLContext];
+}
+
+
+-(void) lockOpenGLContext
+{
+    NSOpenGLContext *glContext = [self openGLContext];
+    NSAssert( glContext, @"FATAL: could not get openGL context");
+    
+    [glContext makeCurrentContext];
+    CGLLockContext([glContext CGLContextObj]);
+}
+
+-(void) unlockOpenGLContext
+{
+    NSOpenGLContext *glContext = [self openGLContext];
+    NSAssert( glContext, @"FATAL: could not get openGL context");
+    
+    CGLUnlockContext([glContext CGLContextObj]);
+}
+
+// Find or make a fence that is ready to use.
+-(CCGLFence *)getReadyFence
+{
+    // First checkf oldest (first in the array) fence is ready again.
+    CCGLFence *fence = _fences.firstObject;;
+    if(fence.isReady){
+        // Remove the fence so it can be inserted at the end of the queue again.
+        [_fences removeObjectAtIndex:0];
+        return fence;
+    } else {
+        // No existing fences ready. Make a new one.
+        return [[CCGLFence alloc] init];
+    }
+}
+
+-(void)addFrameCompletionHandler:(dispatch_block_t)handler
+{
+    if(_fences == nil){
+        _fences = [NSMutableArray arrayWithObject:[[CCGLFence alloc] init]];
+    }
+    
+    CCGLFence *fence = _fences.lastObject;
+    if(!fence.isReady){
+        fence = [self getReadyFence];
+        [_fences addObject:fence];
+    }
+    
+    [fence.handlers addObject:handler];
+}
+
+-(void)beginFrame
+{
+    [self lockOpenGLContext];
+}
+
+-(void)presentFrame
+{
+    {
+        CCGLFence *fence = _fences.lastObject;
+        if(fence.isReady){
+            // If the fence is ready to be added, insert a sync point for it.
+            [fence insertFence];
+        }
+    }
+    
+    [self.openGLContext flushBuffer];
+    
+    // Check the fences for completion.
+    for(CCGLFence *fence in _fences){
+        if(fence.isCompleted){
+            for(dispatch_block_t handler in fence.handlers) handler();
+            [fence.handlers removeAllObjects];
+        } else {
+            break;
+        }
+    }
+    
+    [self unlockOpenGLContext];
+}
+
+-(GLuint)fbo
+{
+    return 0;
+}
+
+- (void) dealloc
+{
+    CCLOGINFO(@"cocos2d: deallocing %@", self);
+}
+
+#pragma mark CCGLView - Mouse Delegate
+
+- (void)mouseDown:(NSEvent *)theEvent
+{
+    // dispatch mouse to responder manager
+    [CCDirector bindDirector:_director];
```

Comment:
wrt the discussion: in what situation would you wish to have more than one CCView in an application? The concept of multiple directors makes sense in this situation if you have more than one view and wish to display multiple scenes simultaneously, but I can't seem to envision a use case that can't be achieved with render targets.

As far as testability goes, I feel this issue is due to the dependencies of discrete objects within cocos2d on a director rather than the fact that it exists as a singleton. I'd like to point out that I'm in no way attached to the implementation of the director as a singleton.

I'm totally for moving the platform specific code out of the director instances, this would really help differentiate between platform / GL initialization setup and what is relevant for cocos configuration. 

I've started with a pull request to make this clearer with regards to https://github.com/spritebuilder/SpriteBuilder/issues/912, but it has significant crossover with your changes here and I don't want to complicate your efforts (esp with the impending v4 deadline) so I will wait for this to go in.